static/app/components/charts/miniBarChart.tsx

```diff
@@ -224,6 +224,45 @@ export function getYAxisMaxFn(height: number) {
   };
 }
 
+function groupDataByName(series: BarChartSeries[]): Map<string | number, number[]> {
+  const groupedData = new Map<string | number, number[]>();
+  series.forEach(serie => {
+    serie.data.forEach(({name, value}) => {
+      if (!groupedData.has(name)) {
+        groupedData.set(name, []);
+      }
+      groupedData.get(name)?.push(value);
+    });
+  });
+
+  return groupedData;
+}
+
+// Adds rounded corners to a single bar or the last item of a stacked bar based on the datapoints
+function updateDataItemBorderRadius(
+  groupedData: Map<string | number, number[]>,
+  data: SeriesDataUnit[],
+  serieIndex: number
+) {
+  return data.map(dataItem => {
+    const datapointsByName = groupedData.get(dataItem.name) ?? [];
+    const allAreZero = datapointsByName.every(value => value === 0);
+    const lastNonZeroIndex = datapointsByName.map(value => value > 0).lastIndexOf(true);
+    const isLastStackedItem = lastNonZeroIndex === serieIndex;
+
+    if (allAreZero || isLastStackedItem) {
+      return {
+        ...dataItem,
+        itemStyle: {
+          borderRadius: [1, 1, 0, 0],
+        },
+      };
+    }
+
+    return dataItem;
+  });
+}
```

Comment:
using Object.groupBy can reduce the code in groupDataByName

```suggestion
function groupDataByName(
  series: BarChartSeries[]
): Record<string, SeriesDataUnit[] | undefined> {
  const allData = series.flatMap(serie => serie.data);
  return Object.groupBy<string, SeriesDataUnit>(allData, ({name}) => String(name));
}

// Adds rounded corners to a single bar or the last item of a stacked bar based on the datapoints
function updateDataItemBorderRadius(
  groupedData: Record<string, SeriesDataUnit[] | undefined>,
  data: SeriesDataUnit[],
  serieIndex: number
) {
  return data.map(dataItem => {
    const datapointsByName = groupedData[dataItem.name] ?? [];
    const allAreZero = datapointsByName.every(value => value.value === 0);
    const lastNonZeroIndex = datapointsByName
      .map(value => value.value > 0)
      .lastIndexOf(true);
    const isLastStackedItem = lastNonZeroIndex === serieIndex;

    if (allAreZero || isLastStackedItem) {
      return {
        ...dataItem,
        itemStyle: {
          borderRadius: [1, 1, 0, 0],
        },
      };
    }

    return dataItem;
  });
}
```

Reply:
Nice! I don't remember using `Object.groupBy` before. TIL :)